File: code/go/day37/day37.go
package day37

func partition(data *[]int, length, start, end int) int {
	pivot := start
	index := pivot + 1

	for i := index; i <= end; i++ {
		if (*data)[i] < (*data)[pivot] {
			(*data)[i], (*data)[index] = (*data)[index], (*data)[i]
			index++
		}
	}
	(*data)[pivot], (*data)[index - 1] = (*data)[index - 1], (*data)[pivot]
	return index - 1
}

func QuickSort(data *[]int, length, start, end int) {
	if start == end {
		return
	}
	index := partition(data, length, start, end)

	if index > start {
		QuickSort(data, length, start, index - 1)
	}
	if index < end {
		QuickSort(data, length, index + 1, end)
	}
}

File: code/python/day37/day37.py
"""
3, 5, 2, 1, 8, 6
选取1作为基准: 3, 5, 2, 6, 8, 1
"""
import unittest
import random


def swap(l, index1, index2):
    l[index1], l[index2] = l[index2], l[index1]


def partition(numbers, start, end):
    parviot = random.randint(start, end-1)
    small = start - 1
    swap(numbers, parviot, end)

    for i in range(end-start+1):
        index = start+i
        if numbers[index] < numbers[end]:
            small += 1
            if index != small:
                swap(numbers, index, small)

    small += 1
    swap(numbers, small, end)

    return small


def quick_sort(numbers, start, end):
    if start == end:
        return

    privot = partition(numbers, start, end)
    if privot > start:
        quick_sort(numbers, start, privot-1)
    if end > privot:
        quick_sort(numbers, privot+1, end)

    return numbers



class TestQuickSort(unittest.TestCase):

    def test_sort(self):
        self.assertEqual(quick_sort([3, 5, 2, 1, 8, 6], 0, 5), [1, 2, 3, 5, 6, 8])

if __name__ == '__main__':
    unittest.main()
